src/main/java/org/wildfly/discovery/Discovery.java

```diff
@@ -242,7 +284,11 @@ public ServiceURL pollService() {
         }
 
         public ServiceURL takeService() throws InterruptedException {
-            await();
+            if (timeout > 0) {
```

Comment:
This doesn't seem correct.  It will reapply the original timeout on every discovery request.  So if your timeout is 5 seconds, and you get 15 results at 4-second intervals, your program will block for a minute - surely not the correct intention.

It might be better to drop the timeout parameter on the discover method and add a `takeService` variant with timeout instead.  This way, the caller can keep track of an overall timeout (with a diminishing `remaining` value) and allow discovery to run until that time.

Reply:
How so? My understanding is that `queue.poll` will return as soon as there is a result or give up if the timeout is reached.

Reply:
The `poll` method will return immediately either way.  But the way this is written, every call to `takeService` will have the same overall timeout, which is probably not what is wanted.  It's the overall request that has a timeout, not the individual discovery request; all the timeouts of all the steps involved should end at the same moment if possible.